src/relay/pass/ad.cc

```diff
@@ -0,0 +1,180 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file ad.cc
+ * \brief API for Automatic Differentiation for the Relay IR.
+ */
+
+#include <tvm/lowered_func.h>
+#include <tvm/operation.h>
+#include <tvm/relay/ad.h>
+#include <tvm/relay/builder.h>
+#include <tvm/relay/expr_functor.h>
+#include <tvm/relay/pass.h>
+#include "let_list.h"
+#include "../ir/type_functor.h"
+
+namespace tvm {
+namespace relay {
+
+using namespace tvm::runtime;
+
+Type WithGradientType(const Type& t) {
+  // TODO(M.K.): stricter checking
+  auto ty = t.as<FuncTypeNode>();
+  CHECK(ty) << "input should be a function";
+  return FuncTypeNode::make(ty->arg_types,
+                            TupleTypeNode::make({
+                              ty->ret_type,
+                              TupleTypeNode::make(ty->arg_types)}), {}, {});
+}
+
+//! \brief if the expression is a GlobalVar, transform to it's expression.
+Expr DeGlobal(const Module& mod, const Expr& e) {
```

Comment:
Are nested globals handled appropriately?

Reply:
Ah, I see. Perhaps a clarifying comment that the expression is expected to be a function might be useful